File: lc_py/lcPy/dp/MaxProduct152.py
from typing import List
class MaxProduct:
    def maxProduct(self,nums: List[int]) -> int:
        n = len(nums)
        # 初始化动态规划数组
        dp_max = [0] * n
        dp_min = [0] * n
        # 设置初始值
        dp_max[0] = dp_min[0] = nums[0]
        max_product = nums[0]
        
        # 遍历数组，计算每个位置的最大和最小乘积
        for i in range(1, n):
            # 计算当前位置的最大乘积
            dp_max[i] = max(nums[i], dp_max[i-1] * nums[i], dp_min[i-1] * nums[i])
            # 计算当前位置的最小乘积
            dp_min[i] = min(nums[i], dp_max[i-1] * nums[i], dp_min[i-1] * nums[i])
            # 更新全局最大乘积
            max_product = max(max_product, dp_max[i])
            
        return max_product

File: lc_go/lcGo/dp/maxProduct152.go
package main

import "fmt"

// maxProduct 计算数组中连续子数组的最大乘积
// 使用动态规划，维护两个数组：
// dpMax[i] 表示以 nums[i] 结尾的子数组的最大乘积
// dpMin[i] 表示以 nums[i] 结尾的子数组的最小乘积
func maxProduct(nums []int)(maxProducts int) {
	n := len(nums)
	// 初始化 dp 数组
	dpMax := make([]int,n)
	dpMin := make([]int,n)
	
	// 初始化第一个元素
	dpMax[0] = nums[0]
	dpMin[0] = nums[0]
	maxProducts = nums[0]

	// 遍历数组，计算每个位置的最大和最小乘积
	for i := 1; i < n; i++ {
		// 当前元素可能是新的子数组起点，或者与前面的子数组相乘
		// 需要考虑正负号的影响，所以同时维护最大和最小值
		dpMax[i] = max(nums[i], max(dpMax[i-1]*nums[i], dpMin[i-1]*nums[i]))
		dpMin[i] = min(nums[i], min(dpMax[i-1]*nums[i], dpMin[i-1]*nums[i]))
		// 更新全局最大值
		maxProducts = max(maxProducts, dpMax[i])
	}
	return 
}

func main(){
	nums := []int{-2,3,-4}
	res := maxProduct(nums)
	fmt.Println(res)
}